File: KubiTest/ViewController.m
//
//  ViewController.m
//  KubiTest
//
//  Created by Victor Nouvellet on 19/01/2017.
//  Copyright © 2017 Victor Nouvellet Inc. All rights reserved.
//

#import "ViewController.h"
#import "GCDWebServer.h"
#import "GCDWebServerDataResponse.h"

@interface ViewController ()

@property (nullable, nonatomic, weak) RRDeviceSDK* sdk;
@property (nonatomic, strong) UIAlertController* connectionAlert;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *scanIndicator;
@property (weak, nonatomic) IBOutlet UIButton *scanButton;
@property (weak, nonatomic) IBOutlet UIView *connectionView;
@property (weak, nonatomic) IBOutlet UIView *controlView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *topConstraint;

@property BOOL isScanning;
@property GCDWebServer* webServer;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.sdk = [RRDeviceSDK deviceSdk];
    self.sdk.delegate = self;
    [self updateControlButtons];
    
    [self startScanning];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)scanButtonPressed:(id)sender {
    [self updateControlButtons];
    [self startScanning];
}

- (IBAction)disconnectButtonPressed:(id)sender {
    [self.sdk disconnectDevice];
}

- (IBAction)shareButtonPressed:(id)sender {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    if (kubi != nil) {
        // shareControlWithAppData method seems to be broken
//        [self.sdk shareControlWithAppData:nil success:^{
//            NSLog(@"SUCCESS");
//        } fail:^(NSError * _Nullable error) {
//            NSLog(@"FAIL");
//        }];
        NSLog(@"Identifier : %@ \n Type : %@", kubi.identifier, kubi.type);
    }
    
    // Create server
    self.webServer = [[GCDWebServer alloc] init];
    
    // Add a handler to respond to GET requests on any URL
    __weak __typeof__(self) weakSelf = self;
    [self.webServer addDefaultHandlerForMethod:@"GET"
                                  requestClass:[GCDWebServerRequest class]
                                  processBlock:^GCDWebServerResponse *(GCDWebServerRequest* request) {
                                      
                                      NSLog(@"New client!");
                                      
                                      dispatch_async(dispatch_get_main_queue(), ^{
                                          [weakSelf handleRequest:request];
                                      });
                                      
                                      return [GCDWebServerDataResponse responseWithHTML:@"<html><body><p>Hello World</p></body></html>"];
                                      
                                  }];
    
    // Start server on port 80
    [self.webServer startWithPort:80 bonjourName:@"KUBI"];
    NSLog(@"Visit %@ or %@ in your web browser", self.webServer.serverURL, self.webServer.bonjourServerURL);
}

-(void)deviceSDK:(RRDeviceSDK *)deviceSDK didChangeConnectionState:(RRDeviceConnectionState)connectionState {
    NSLog(@"device SDK didChangeConnectionState");
    [self updateControlButtons];
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    switch (connectionState) {
        case RRDeviceConnectionStateConnected:
            [self.connectionAlert dismissViewControllerAnimated:true completion:nil];
            
            [kubi setPanEnabled:true error:nil];
            [kubi setTiltEnabled:true error:nil];
            
            break;
        case RRDeviceConnectionStateConnecting:
            [self.connectionAlert dismissViewControllerAnimated:true completion:nil];
            break;
        case RRDeviceConnectionStateDisconnected:
            break;
        case RRDeviceConnectionStateConnectionLost:
            break;
            
        default:
            break;
    }
}

-(void)deviceSDK:(RRDeviceSDK *)deviceSDK didUpdateDeviceList:(NSArray *)deviceList {
    NSLog(@"device SDK didUpdateDeviceList");
    
    if (self.sdk.deviceConnectionState == RRDeviceConnectionStateConnecting || self.sdk.deviceConnectionState == RRDeviceConnectionStateConnected) {
        [self stopScanning];
        return;
    }
    
    self.connectionAlert = [UIAlertController alertControllerWithTitle:@"Choose your device"
                                                                       message:@"Here is the list of available devices"
                                                                preferredStyle:UIAlertControllerStyleActionSheet];
    
    for (RRDevice* device in deviceList) {
        UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:device.name style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action) {
                                                                  [self stopScanning];
                                                                  [self.sdk connectDevice:device];
                                                              }];
        [self.connectionAlert addAction:defaultAction];
    }
    
    UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"Wait..." style:UIAlertActionStyleDestructive
                                                          handler:^(UIAlertAction * action) {
                                                              
                                                          }];
    [self.connectionAlert addAction:defaultAction];
    
    UIAlertAction* cancelAction = [UIAlertAction actionWithTitle:@"Stop scanning" style:UIAlertActionStyleCancel
                                                          handler:^(UIAlertAction * action) {
                                                              [self stopScanning];
                                                          }];
    [self.connectionAlert addAction:cancelAction];
    
    
    if (!self.connectionAlert.isBeingPresented && self.isScanning == true) {
        [self presentViewController:self.connectionAlert animated:YES completion:nil];
    }
}

#pragma mark - Scan functions

-(void)startScanning {
    self.isScanning = true;
    self.scanButton.enabled = false;
    [self.scanIndicator startAnimating];
    [self.sdk startScan];
}

-(void)stopScanning {
    [self.sdk endScan];
    [self.scanIndicator stopAnimating];
    self.scanButton.enabled = true;
    self.isScanning = false;
}

#pragma mark - Remote control

-(void)updateControlButtons {
    if (self.sdk.deviceConnectionState == RRDeviceConnectionStateConnected) {
        //Activate button visibility with animation
        [UIView animateWithDuration:1.5
                              delay:0
             usingSpringWithDamping:0.5
              initialSpringVelocity:0.5
                            options:UIViewAnimationOptionCurveEaseInOut
                         animations:^{
                             //What to animate
                             self.topConstraint.constant = 20;
                             [self.view layoutIfNeeded];
                         } completion:^(BOOL finished) {
                             //What to do after completion
                         }];
        self.controlView.hidden = false;
    }   else  {
        //Hide control buttons
        [UIView animateWithDuration:1.5
                              delay:0
             usingSpringWithDamping:0.5
              initialSpringVelocity:0.5
                            options:UIViewAnimationOptionCurveEaseInOut
                         animations:^{
                             //What to animate
                             self.topConstraint.constant = 250;
                             [self.view layoutIfNeeded];
                         } completion:^(BOOL finished) {
                             //What to do after completion
                         }];
        self.controlView.hidden = true;
    }
}

- (IBAction)tiltUpButtonPressed:(id)sender {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:0]
                           atPanSpeed:[NSNumber numberWithDouble:0]
                         andTiltDelta:[NSNumber numberWithDouble:20]
                          atTiltSpeed:[NSNumber numberWithDouble:150]
                                error:nil];
}

- (IBAction)tiltDownButtonPressed:(id)sender {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:0]
                           atPanSpeed:[NSNumber numberWithDouble:0]
                         andTiltDelta:[NSNumber numberWithDouble:-20]
                          atTiltSpeed:[NSNumber numberWithDouble:150]
                                error:nil];
}

- (IBAction)panLeftButtonPressed:(id)sender {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:-20]
                           atPanSpeed:[NSNumber numberWithDouble:150]
                         andTiltDelta:[NSNumber numberWithDouble:0]
                          atTiltSpeed:[NSNumber numberWithDouble:0]
                                error:nil];
}

- (IBAction)panRightButtonPressed:(id)sender {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:20]
                           atPanSpeed:[NSNumber numberWithDouble:150]
                         andTiltDelta:[NSNumber numberWithDouble:0]
                          atTiltSpeed:[NSNumber numberWithDouble:0]
                                error:nil];
}

#pragma mark Requests handling

-(void)handleRequest:(GCDWebServerRequest*) request {
    RRKubi* kubi = (RRKubi*)self.sdk.connectedDevice;
    if(kubi != nil && [request.URL.relativePath isEqualToString:@"/"]) {
        if (request.query[@"panDelta"] != nil && request.query[@"panSpeed"] != nil && request.query[@"tiltDelta"] != nil && request.query[@"tiltSpeed"] != nil) {
            double panDelta = [request.query[@"panDelta"] doubleValue] ?: 0;
            double panSpeed = [request.query[@"panSpeed"] doubleValue] ?: 0;
            double tiltDelta = [request.query[@"tiltDelta"] doubleValue] ?: 0;
            double tiltSpeed = [request.query[@"tiltSpeed"] doubleValue] ?: 0;
            [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:panDelta]
                                   atPanSpeed:[NSNumber numberWithDouble:panSpeed]
                                 andTiltDelta:[NSNumber numberWithDouble:tiltDelta]
                                  atTiltSpeed:[NSNumber numberWithDouble:tiltSpeed]
                                        error:nil];
        } else {
            [kubi incrementalMoveWithPanDelta:[NSNumber numberWithDouble:5]
                           atPanSpeed:[NSNumber numberWithDouble:150]
                         andTiltDelta:[NSNumber numberWithDouble:0]
                          atTiltSpeed:[NSNumber numberWithDouble:0]
                                error:nil];
        }
    }
}

@end

File: KubiTest/ViewController.swift
//
//  ViewController.swift
//  KubiTest
//
//  Created by Victor Nouvellet on 1/23/17.
//  Copyright © 2017 Victor Nouvellet Inc. All rights reserved.
//

import UIKit
import GCDWebServer
import KubiDeviceSDK

class ViewController: UIViewController {
    
    @IBOutlet weak var scanIndicator: UIActivityIndicatorView?
    @IBOutlet weak var scanButton: UIButton?
    @IBOutlet weak var connectionView: UIView?
    @IBOutlet weak var controlView: UIView?
    @IBOutlet weak var topConstraint: NSLayoutConstraint?
    
    var connectionAlert: UIAlertController = UIAlertController()
    var isScanning: Bool = false
    var kubiManager: KubiManager = KubiManager.sharedInstance
    var webServer: GCDWebServer?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.configureKubiManagerCallbacks()
        
        self.updateControlButtons()
        _ = self.startScanning()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    @IBAction func scanButtonPressed(sender: UIButton) {
        self.updateControlButtons()
        _ = self.startScanning()
    }
    
    @IBAction func disconnectButtonPressed(sender: UIButton) {
        self.kubiManager.disconnectDevice()
    }
    
    @IBAction func shareButtonPressed(sender: UIButton) {
        if let kubi = self.kubiManager.connectedDevice as? RRKubi {
            // shareControlWithAppData method seems to be broken
            //        [self.sdk shareControlWithAppData:nil success:^{
            //            NSLog(@"SUCCESS");
            //        } fail:^(NSError * _Nullable error) {
            //            NSLog(@"FAIL");
            //        }];
            print("Identifier : %@ \n Type : %@", kubi.identifier, kubi.type)
        }
        
        // Create server
        self.webServer = GCDWebServer()
        
        // Add a handler to respond to GET requests on any URL
        self.webServer?.addDefaultHandler(forMethod: "GET",
                                          request: GCDWebServerRequest.self,
                                          asyncProcessBlock: { (request: GCDWebServerRequest?, completionBlock: GCDWebServerCompletionBlock?) in
                                            print("New client!")
                                            self.handleRequest(request: request, completionBlock: completionBlock)
        })
        
        // Start server on port 80
        self.webServer?.start(withPort: 80, bonjourName: "KUBI-Server")
        
        print("Visit \(self.webServer?.serverURL) or \(self.webServer?.bonjourServerURL) in your web browser")

    }
    
    // MARK: Private methods
    
    private func configureKubiManagerCallbacks() {
        self.kubiManager.deviceDidUpdateDeviceListCallback = { (deviceList: [Any]) in
            
            print("device SDK didUpdateDeviceList")
            
            if self.kubiManager.deviceConnectionState == .connecting || self.kubiManager.deviceConnectionState == .connecting {
                self.stopScanning()
                return
            }
            
            self.connectionAlert = UIAlertController(title: "Choose your device", message: "Here is the list of available devices", preferredStyle: .actionSheet)
            
            for device in deviceList {
                guard let safeDevice = device as? RRKubi else {
                    continue
                }
                let defaultAction: UIAlertAction = UIAlertAction(title: safeDevice.name(), style: .default, handler: { (action: UIAlertAction) in
                    
                    self.stopScanning()
                    self.kubiManager.kubiSdk.connect(safeDevice)
                })
                self.connectionAlert.addAction(defaultAction)
            }
            
            let defaultAction: UIAlertAction = UIAlertAction(title: "Wait...", style: .destructive, handler: nil)
            self.connectionAlert.addAction(defaultAction)
            
            let cancelAction: UIAlertAction = UIAlertAction(title: "Stop scanning", style: .destructive, handler: {
                (action: UIAlertAction) in
                self.stopScanning()
            })
            self.connectionAlert.addAction(cancelAction)
            
            if !self.connectionAlert.isBeingPresented && self.isScanning == true {
                self.present(self.connectionAlert, animated: true, completion: nil)
            }
        }
        
        self.kubiManager.deviceDidChangeConnectionCallback = { (connectionState: RRDeviceConnectionState) in
            
            print("device SDK didChangeConnectionState")
            
            self.updateControlButtons()
            switch connectionState {
            case .connected:
                self.connectionAlert.dismiss(animated: true, completion: nil)
                if let kubi = self.kubiManager.connectedDevice as? RRKubi {
                    do {
                        try kubi.setPanEnabled(true)
                        try kubi.setTiltEnabled(true)
                    } catch {
                        print("[WARNING] Tilt or Pan not enabled properly")
                    }
                }
                break
            case .connecting:
                self.connectionAlert.dismiss(animated: true, completion: nil)
                break
            case .disconnected:
                break
            case .connectionLost:
                break
            }
        }
    }
    
    private func updateControlButtons() {
        if self.kubiManager.deviceConnectionState == .connected {
            UIView.animate(withDuration: 1.5,
                           delay: 0,
                           usingSpringWithDamping: 0.5,
                           initialSpringVelocity: 0.5,
                           options: .curveEaseInOut,
                           animations: {
                            self.topConstraint?.constant = 20
                            self.view.layoutIfNeeded()
            }, completion: { (complete: Bool) in
                //What to do after completion
            })
            
            self.controlView?.isHidden = false
        }   else    {
            UIView.animate(withDuration: 1.5,
                           delay: 0,
                           usingSpringWithDamping: 0.5,
                           initialSpringVelocity: 0.5,
                           options: .curveEaseInOut,
                           animations: {
                            self.topConstraint?.constant = 20
                            self.view.layoutIfNeeded()
            }, completion: { (complete: Bool) in
                //What to do after completion
            })
            
            self.controlView?.isHidden = true
        }
    }
    
    // MARK: - Scan functions
    
    func startScanning() -> Bool {
        self.scanButton?.isEnabled = false
        self.scanIndicator?.startAnimating()
        self.isScanning = self.kubiManager.startScan()
        return self.isScanning
    }
    
    func stopScanning() {
        self.kubiManager.endScan()
        self.isScanning = false
        self.scanButton?.isEnabled = false
        self.scanIndicator?.stopAnimating()
    }
    
    // MARK: - Remote Control
    
    @IBAction func tiltUpButtonPressed(sender: UIButton) {
        guard let kubi = self.kubiManager.connectedDevice as? RRKubi
            else {
                print("[Warning] No Kubi connected = No remote control")
                return
        }
        
        do {
            try kubi.incrementalMove(withPanDelta: NSNumber(value: 0),
                             atPanSpeed: NSNumber(value: 0),
                             andTiltDelta: NSNumber(value: 20),
                             atTiltSpeed: NSNumber(value: 150))
        } catch {
            print("Error doing incremental")
        }
    }
    
    @IBAction func tiltDownButtonPressed(sender: UIButton) {
        guard let kubi = self.kubiManager.connectedDevice as? RRKubi
            else {
                print("[Warning] No Kubi connected = No remote control")
                return
        }
        
        do {
            try kubi.incrementalMove(withPanDelta: NSNumber(value: 0),
                                     atPanSpeed: NSNumber(value: 0),
                                     andTiltDelta: NSNumber(value: 20),
                                     atTiltSpeed: NSNumber(value: 150))
        } catch {
            print("Error doing incremental")
        }
    }
    
    @IBAction func panRightButtonPressed(sender: UIButton) {
        guard let kubi = self.kubiManager.connectedDevice as? RRKubi
            else {
                print("[Warning] No Kubi connected = No remote control")
                return
        }
        
        do {
            try kubi.incrementalMove(withPanDelta: NSNumber(value: 0),
                                     atPanSpeed: NSNumber(value: 0),
                                     andTiltDelta: NSNumber(value: 20),
                                     atTiltSpeed: NSNumber(value: 150))
        } catch {
            print("Error doing incremental")
        }
    }
    
    @IBAction func panLeftButtonPressed(sender: UIButton) {
        guard let kubi = self.kubiManager.connectedDevice as? RRKubi
            else {
                print("[Warning] No Kubi connected = No remote control")
                return
        }
        
        do {
            try kubi.incrementalMove(withPanDelta: NSNumber(value: 0),
                                     atPanSpeed: NSNumber(value: 0),
                                     andTiltDelta: NSNumber(value: 20),
                                     atTiltSpeed: NSNumber(value: 150))
        } catch {
            print("Error doing incremental")
        }
    }
    
    // MARK: - Requests handling
    
    func handleRequest(request: GCDWebServerRequest?, completionBlock: GCDWebServerCompletionBlock?) {
        
        guard let device: RRDevice = self.kubiManager.connectedDevice,
            let kubi: RRKubi = device as? RRKubi,
            request?.url.relativePath == "/"
            else {
                print("No connected kubi")
                let responseFail = GCDWebServerResponse()
                completionBlock?(responseFail)
                return
        }
        
        if request?.query["panDelta"] != nil && request?.query["panSpeed"] != nil &&
            request?.query["tiltDelta"] != nil && request?.query["tiltSpeed"] != nil {
            let panDelta: NSNumber = request?.query["panDelta"] as? NSNumber ?? 0
            let panSpeed: NSNumber = request?.query["panSpeed"] as? NSNumber ?? 0
            let tiltDelta: NSNumber = request?.query["tiltDelta"] as? NSNumber ?? 0
            let tiltSpeed: NSNumber = request?.query["tiltSpeed"] as? NSNumber ?? 0
            
            do {
                try kubi.incrementalMove(withPanDelta: panDelta, atPanSpeed: panSpeed, andTiltDelta: tiltDelta, atTiltSpeed: tiltSpeed)
                let responseSuccess = GCDWebServerResponse()
                completionBlock?(responseSuccess)
            } catch {
                print("Error doing incremental")
                let responseFail = GCDWebServerResponse()
                completionBlock?(responseFail)
            }
            
        } else {
            
            let responseFail = GCDWebServerResponse()
            completionBlock?(responseFail)
        }
        
        
    }
}
